### PR TITLE
Set correct int64 flag for petsc +mumps

### PR DESCRIPTION
**Pull request by melven**
Fixes a problem when trying to concretize petsc with +mumps+int64.

### CODE REVIEW

**Review by adamjstewart — APPROVED** [review]
@balay @barrysmith @jedbrown can you review this?

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -227,7 +227,8 @@ class Petsc(Package):
     depends_on('superlu-dist@xsdk-0.2.0+int64', when='@xsdk-0.2.0+superlu-dist+mpi+int64')
     depends_on('superlu-dist@develop~int64', when='@develop+superlu-dist+mpi~int64')
     depends_on('superlu-dist@develop+int64', when='@develop+superlu-dist+mpi+int64')
-    depends_on('mumps+mpi~int64', when='+mumps')
```
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py:230]: Current code is correct. `petsc+int64` uses `mumps~int64` which includes 64int interfaces.

`mumps+int64` I believe compiles everything with  `-i8` which requires ILP64

Comment (melven) [reply var/spack/repos/builtin/packages/petsc/package.py:230]: I got an error during concretize with the old version which is fixed by this:
```
zoel_ml@fe-store01:~/spack (develop)> spack --version
0.15.4-950-afb088376
zoel_ml@fe-store01:~/spack (develop)> spack spec petsc+mumps+suite-sparse+superlu-dist+int64 ^superlu-dist@6.1.1+int64 ^mumps+metis+parmetis+mpi+int64 ^parmetis+int64Input spec
--------------------------------
petsc+int64+mumps+suite-sparse+superlu-dist
    ^mumps+int64+metis+mpi+parmetis
    ^parmetis+int64
    ^superlu-dist@6.1.1+int64

Concretized
--------------------------------
==> Error: An unsatisfiable variant constraint has been detected for spec:

    mumps+int64+metis+mpi+parmetis


while trying to concretize the partial spec:

    petsc+int64+mumps+suite-sparse+superlu-dist
        ^blas
        ^diffutils
            ^iconv
        ^lapack


petsc requires mumps variant ~int64, but spec asked for +int64
```

With this fix:
```
zoel_ml@fe-store01:~/spack (develop)> git checkout petsc_mumps_int64
Switched to branch 'petsc_mumps_int64'
Your branch is up to date with 'melven/petsc_mumps_int64'.
zoel_ml@fe-store01:~/spack (petsc_mumps_int64)> spack --version
0.15.4-948-30d150bfc
zoel_ml@fe-store01:~/spack (petsc_mumps_int64)> spack spec petsc+mumps+suite-sparse+superlu-dist+int64 ^superlu-dist@6.1.1+int64 ^mumps+metis+parmetis+mpi+int64 ^parmetis+int64
Input spec
--------------------------------
petsc+int64+mumps+suite-sparse+superlu-dist
    ^mumps+int64+metis+mpi+parmetis
    ^parmetis+int64
    ^superlu-dist@6.1.1+int64

Concretized
--------------------------------
==> Warning: the attribute "paths" in the "packages" section of the configuration has been deprecated [entry=CommentedMap([('slurm@20-02-0-1 +hwloc', '/usr/')])]
petsc@3.13.4%gcc@10.2.0~X~batch~cgns~complex~cuda~debug+double~exodusii~fftw~giflib+hdf5+hypre+int64~jpeg~knl~libpng~libyaml~memkind+metis~moab~mpfr+mpi+mumps~p4est~random123~saws+shared+suite-sparse+superlu-dist~trilinos~valgrind clanguage=C arch=linux-ubuntu18.04-skylake_avx512
    ^diffutils@3.7%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
        ^libiconv@1.16%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
    ^hdf5@1.10.6%gcc@10.2.0~cxx~debug~fortran~hl~java+mpi+pic+shared~szip~threadsafe api=none arch=linux-ubuntu18.04-skylake_avx512
        ^openmpi@3.1.6%gcc@10.2.0~atomics~cuda~cxx~cxx_exceptions+gpfs~java+legacylaunchers~lustre~memchecker+pmi~singularity~sqlite3+static~thread_multiple+vt+wrapper-rpath fabrics=auto schedulers=slurm arch=linux-ubuntu18.04-skylake_avx512
            ^hwloc@1.11.11%gcc@10.2.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml+pci+shared arch=linux-ubuntu18.04-skylake_avx512
                ^libpciaccess@0.16%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                    ^libtool@2.4.6%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                        ^m4@1.4.18%gcc@10.2.0+sigsegv patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=linux-ubuntu18.04-skylake_avx512
                            ^libsigsegv@2.12%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                    ^pkgconf@1.7.3%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                    ^util-macros@1.19.1%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                ^libxml2@2.9.10%gcc@10.2.0~python arch=linux-ubuntu18.04-skylake_avx512
                    ^xz@5.2.5%gcc@10.2.0~pic arch=linux-ubuntu18.04-skylake_avx512
                    ^zlib@1.2.11%gcc@10.2.0+optimize+pic+shared arch=linux-ubuntu18.04-skylake_avx512
                ^numactl@2.0.12%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                    ^autoconf@2.69%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                        ^perl@5.30.3%gcc@10.2.0+cpanm+shared+threads arch=linux-ubuntu18.04-skylake_avx512
                            ^berkeley-db@18.1.40%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                            ^gdbm@1.18.1%gcc@10.2.0 patches=7f569b9287e9e5a4236636eef32ff3428f0462ec17aea359e0ebee07649f1f16 arch=linux-ubuntu18.04-skylake_avx512
                                ^readline@8.0%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
                                    ^ncurses@6.2%gcc@10.2.0~symlinks+termlib arch=linux-ubuntu18.04-skylake_avx512
                    ^automake@1.16.2%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
            ^slurm@20-02-0-1%gcc@10.2.0~gtk~hdf5+hwloc~mariadb~pmix+readline sysconfdir=PREFIX/etc arch=linux-ubuntu18.04-skylake_avx512
    ^hypre@2.18.2%gcc@10.2.0~complex~debug+int64~internal-superlu~mixedint+mpi~openmp+shared~superlu-dist arch=linux-ubuntu18.04-skylake_avx512
        ^openblas@0.3.10%gcc@10.2.0~consistent_fpcsr~ilp64+pic+shared threads=none arch=linux-ubuntu18.04-skylake_avx512
    ^metis@5.1.0%gcc@10.2.0~gdb+int64~real64+shared build_type=Release patches=4991da938c1d3a1d3dea78e49bbebecba00273f98df2a656e38b83d55b281da1,b1225da886605ea558db7ac08dd8054742ea5afe5ed61ad4d0fe7a495b1270d2 arch=linux-ubuntu18.04-skylake_avx512
        ^cmake@3.18.2%gcc@10.2.0~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu18.04-skylake_avx512
            ^openssl@1.1.1g%gcc@10.2.0+systemcerts arch=linux-ubuntu18.04-skylake_avx512
    ^mumps@5.3.3%gcc@10.2.0+complex+double+float+int64+metis+mpi+parmetis~ptscotch~scotch+shared arch=linux-ubuntu18.04-skylake_avx512
        ^netlib-scalapack@2.1.0%gcc@10.2.0~pic+shared build_type=RelWithDebInfo patches=1c9ce5fee1451a08c2de3cc87f446aeda0b818ebbce4ad0d980ddf2f2a0b2dc4,f2baedde688ffe4c20943c334f580eb298e04d6f35c86b90a1f4e8cb7ae344a2 arch=linux-ubuntu18.04-skylake_avx512
        ^parmetis@4.0.3%gcc@10.2.0~gdb+int64+shared build_type=RelWithDebInfo patches=4f892531eb0a807eb1b82e683a416d3e35154a455274cf9b162fb02054d11a5b,50ed2081bc939269689789942067c58b3e522c269269a430d5d34c00edbc5870,704b84f7c7444d4372cb59cca6e1209df4ef3b033bc4ee3cf50f369bce972a9d arch=linux-ubuntu18.04-skylake_avx512
    ^python@3.8.5%gcc@10.2.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87 arch=linux-ubuntu18.04-skylake_avx512
        ^bzip2@1.0.8%gcc@10.2.0+shared arch=linux-ubuntu18.04-skylake_avx512
        ^expat@2.2.9%gcc@10.2.0+libbsd arch=linux-ubuntu18.04-skylake_avx512
            ^libbsd@0.10.0%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
        ^gettext@0.21%gcc@10.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu18.04-skylake_avx512
            ^tar@1.32%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
        ^libffi@3.3%gcc@10.2.0 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-ubuntu18.04-skylake_avx512
        ^libuuid@1.0.3%gcc@10.2.0 arch=linux-ubuntu18.04-skylake_avx512
        ^sqlite@3.33.0%gcc@10.2.0+column_metadata+fts~functions~rtree arch=linux-ubuntu18.04-skylake_avx512
    ^suite-sparse@5.7.2%gcc@10.2.0~cuda~openmp+pic~tbb arch=linux-ubuntu18.04-skylake_avx512
    ^superlu-dist@6.1.1%gcc@10.2.0+int64~openmp+shared build_type=RelWithDebInfo arch=linux-ubuntu18.04-skylake_avx512
```

The variant with the fix compiled fine - I didn't test the binaries, yet.

So if the old code is correct, what should I do differently?

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:230]: Do not use  `+int64` in `^mumps+metis+parmetis+mpi+int64`

Also you do not need `^parmetis+int64`

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:230]: Probably best if mumps variant is changed from `int64` to `ilp64` to avoid this confusion

**Comment by melven** [comment]
Not a fix - old code is correct.

**Pull request closed without merging**